File: packages/core/src/protocol/push.ts
import type { IObjectStore } from "../interfaces.js"
import { nextTimestamp } from "../timestamp.js"
import type { StoredDocument, PushResult } from "./types.js"
import { computeHash } from "./hash.js"
import { computeTimestamps } from "./timestamps.js"

const DOCUMENT_VERSION = 1

/**
 * Push a new full document.
 * - Compares baseHash with current document hash
 * - Match -> accept, compute timestamp diffs, store
 * - Mismatch -> reject with hash_mismatch
 * - baseHash: null for first push (no existing document expected)
 *
 * @param store - Storage backend
 * @param documentKey - Unique key identifying this document in the store
 * @param newData - The full document data to store
 * @param baseHash - Hash of the document the client last saw (null for first push)
 * @param author - Optional author identity for provenance tracking
 */
export async function push(
  store: IObjectStore,
  documentKey: string,
  newData: Record<string, unknown>,
  baseHash: string | null,
  author?: { pubkey: string; signature: string },
): Promise<PushResult> {
  const raw = await store.getString(documentKey)

  let oldData: Record<string, unknown> | null = null
  let oldTimestamps = null
  let currentHash = ""

  if (raw) {
    const existing = JSON.parse(raw) as StoredDocument
    oldData = existing.data
    oldTimestamps = existing.timestamps
    currentHash = existing.hash
  }

  // Hash check
  if (baseHash === null) {
    // First push: document must not exist
    if (raw) {
      return { ok: false, error: "hash_mismatch" }
    }
  } else {
    if (baseHash !== currentHash) {
      return { ok: false, error: "hash_mismatch" }
    }
  }

  const now = nextTimestamp()
  const newHash = await computeHash(newData)
  const timestamps = computeTimestamps(oldData, newData, oldTimestamps, now)

  const doc: StoredDocument = {
    v: DOCUMENT_VERSION,
    data: newData,
    timestamps,
    hash: newHash,
    ...(author && { authorPubkey: author.pubkey, authorSignature: author.signature }),
  }

  await store.put(documentKey, JSON.stringify(doc), { contentType: "application/json" })

  return { ok: true, hash: newHash, timestamp: now }
}

File: servers/python/satellite_server/protocol/push.py
"""Push operation for the Satellite sync protocol."""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

from satellite_server.interfaces import IObjectStore
from satellite_server.timestamp import next_timestamp
from satellite_server.constants import ERROR_HASH_MISMATCH, CONTENT_TYPE_JSON
from satellite_server.protocol.types import (
    StoredDocument,
    PushSuccess,
    PushConflict,
    PushResult,
    DOCUMENT_VERSION,
)
from satellite_server.protocol.hash import compute_hash
from satellite_server.protocol.timestamps import compute_timestamps


@dataclass
class Author:
    """Author identity for provenance tracking."""

    pubkey: str
    signature: str


async def push(
    store: IObjectStore,
    document_key: str,
    new_data: dict[str, Any],
    base_hash: str | None,
    author: Author | None = None,
    skip_timestamps: bool = False,
) -> PushResult:
    """Push a new full document.

    - Compares base_hash with current document hash
    - Match -> accept, compute timestamp diffs, store
    - Mismatch -> reject with hash_mismatch
    - base_hash: None for first push (no existing document expected)
    """
    raw = await store.get_string(document_key)

    old_data: dict[str, Any] | None = None
    old_timestamps = None
    current_hash = ""

    if raw:
        existing = json.loads(raw)
        old_data = existing["data"]
        old_timestamps = existing["timestamps"]
        current_hash = existing["hash"]

    # Hash check
    if base_hash is None:
        if raw:
            return PushConflict(error=ERROR_HASH_MISMATCH)
    else:
        if base_hash != current_hash:
            return PushConflict(error=ERROR_HASH_MISMATCH)

    now = next_timestamp()
    new_hash = compute_hash(new_data)
    timestamps = {} if skip_timestamps else compute_timestamps(old_data, new_data, old_timestamps, now)

    doc: dict[str, Any] = {
        "v": DOCUMENT_VERSION,
        "data": new_data,
        "timestamps": timestamps,
        "hash": new_hash,
    }
    if author:
        doc["authorPubkey"] = author.pubkey
        doc["authorSignature"] = author.signature

    await store.put(document_key, json.dumps(doc), content_type=CONTENT_TYPE_JSON)

    return PushSuccess(hash=new_hash, timestamp=now)
